### DISCUSSION TITLE
Why is cpu usage not fully used when my query nodes are busy working as search latency shows very slowly?

### DISCUSSION THREAD

**Discussion by xiaojuan8** [milvus-io/milvus#22169] [Q&A and General discussion]
Hi, I found that cpu is not fully used, but my query task are not efficiently running on query nodes as the search latency is very high. I dont't know why.

![perform](https://user-images.githubusercontent.com/74544828/218659374-4be133d4-872a-481f-b8fa-d9e9d2145cdd.jpg)
![cpu](https://user-images.githubusercontent.com/74544828/218659381-3e50a35e-4ddd-4e1f-b815-5adae627b781.png)

**Comment by yhmo** [comment]
How many cpu cores are assigned to each query node?
The "Runtime" is the stats of the entire cluster, you can switch to "Query Node" panel.

**Reply by xiaojuan8** [reply]
Hi, thanks. I find it.
![image](https://user-images.githubusercontent.com/74544828/218691201-877d3add-23a4-4340-a885-30c127dcf2d2.png)

**Reply by xiaojuan8** [reply]
> How many cpu cores are assigned to each query node? The "Runtime" is the stats of the entire cluster, you can switch to "Query Node" panel.

actually, I did not specify cpu resource request for querynodes as default.
![image](https://user-images.githubusercontent.com/74544828/218692266-e70050ae-226c-4b34-bbd2-346776f2b304.png)

**Comment by yanliang567** [comment]
a few quick questons:
how many vectors do you have in the collection? what is the dim? what is the index params and search params? what is the search concurrent number?
could you please also share the screenshot of Search in Queue Latency in QueryNode?

**Reply by xiaojuan8** [reply]
1. 100 million vectors in 1 collection
2. dimension=32
3. search_params = {"metric_type": "L2", "params": {"nprobe": 10}, "offset": 5}
4. I have 500 process in python, so concurrent number is 500
5. 
![image](https://user-images.githubusercontent.com/74544828/218717702-b285a5cd-1c3a-4a8b-82e0-03c87d46cd4f.png)

**Reply by yanliang567** [reply]
from the sceenshot above,we can see
1. it costs about 2-4 seconds it search queue, which is too much. So how did you deploy milvus, did you change any milvus configurations when deploying?
2. it costs 500+ms in segcore request, which indicates there might be brute force search. Could you please attach the etcd backup for investigation？ Check [this readme](https://github.com/milvus-io/birdwatcher) for details about how to backup etcd with birdwatcher

**Reply by xiaojuan8** [reply]
I only change replicas to 6 and request cpus to 16 while deploying. OK, I will try to backup etcd and attach it here later.

**Reply by xiaojuan8** [reply]
seems to be some connection error when backup etcd
![image](https://user-images.githubusercontent.com/74544828/218990014-d3abe6d7-de57-4c31-8791-160929744634.png)

**Reply by xiaojuan8** [reply]
I connect etcd node by ip:port from remote client. backup procedure seems to collect infos from other nodes, which should work among cluster nodes.

**Reply by yhmo** [reply]
The error "failed to fetch metrics for xxx" can be ignored. If you saw the message "backup for prefix done, stored in file: xxxx.bak.gz", the file has been created, just send us the file.

**Reply by xiaojuan8** [reply]
Ok, done.
[bw_etcd_ALL.230216-105431.bak.zip](https://github.com/milvus-io/milvus/files/10751786/bw_etcd_ALL.230216-105431.bak.zip)

**Reply by xiaojuan8** [reply]
BTW, how to load or parse this backup file?

**Reply by yhmo** [reply]
The file name should be "bw_etcd_ALL.230216-105431.bak.gz", not "bw_etcd_ALL.230216-105431.bak.zip". Did you unzip it manually?

**Reply by xiaojuan8** [reply]
yes, I gunzip it manually and zip it instead, because I tried to drag the .gz file here, .gz format is not supported to be attached

**Reply by xiaojuan8** [reply]
[bw_etcd_ALL.230216-144048.bak.gz](https://github.com/milvus-io/milvus/files/10752971/bw_etcd_ALL.230216-144048.bak.gz)

**Reply by xiaojuan8** [reply]
> [bw_etcd_ALL.230216-144048.bak.gz](https://github.com/milvus-io/milvus/files/10752971/bw_etcd_ALL.230216-144048.bak.gz)

I tried again, the backup file is uploaded successfully.

**Reply by yhmo** [reply]
The birdwatcher can extract the backup file, it also can load/parse the backup file.

![Screenshot from 2023-02-16 15-05-18](https://user-images.githubusercontent.com/2282099/219292503-cc58007e-2411-49f8-a29d-b0f74145bbd2.png)

**Reply by xiaojuan8** [reply]
Nice!

**Comment by yhmo** [comment]
Sorry, I am wrong on "Estimate CPU usage". This is an estimated value calculated by `nq*segment_num*cpuRatio`, not a runtime value.

The runtime value is displayed in the "Runtime" panel, the "CPU usage".

**Comment by congqixia** [comment]
From the `Search Segment latency` & `Segcore request latency` metrics, it appears that brute force search occurs in your case.
Maybe you could check growing segment number on your query nodes.
The latest dashboard can separate Growing/Sealed segment number for `Segment Loaded Number` metrics.

**Reply by xiaojuan8** [reply]
When does brute-force searching happen? Does happen on growing segment? 
Does it mean too many growing segments leads to brute-force searching?
What would be the reason for too many growing segments? I did not change any config related to segment info.

**Reply by yanliang567** [reply]
Generally, Milvus will automatically seal a segment and build index for it when the segment size meet the threshold. But there will be growing segments if
1. keep inserting 
2. or insert stopped, but there is a small segment that the size does not meet the threshold. In this case, Milvus will seal it after 10 mins or 24 hours depending on how small it is.
One more question, which version of Milvus are you running?

**Reply by xiaojuan8** [reply]
Actually, I have finished inserting for a long time(more than 24 hours). Furthermore, One growing segment(less than 512M) only leads to brute-force searching on a single node theoretically, but there seems to several nodes perform searching very slowly(1~2 second).

My milvus version is v2.2.0

**Reply by yanliang567** [reply]
I suggest you upgrade to v2.2.3, which fixed the issue that not seal small segments even after 24 hours.

**Reply by xiaojuan8** [reply]
> I suggest you upgrade to v2.2.3, which fixed the issue that not seal small segments even after 24 hours.

I am still confused with this issue, only one growing segment not sealed(not indexed) exist in my collection, but several querynodes performs searching slowly at second-level. As mentioned above, only one growing segment which does not meet the threshold to be sealed, there should be only one querynode perform brute-force seaching on the growing segment.

BTW, I will try milvus v2.2.3, thanks!

**Reply by congqixia** [reply]
Just checked your backup file. It looks like all segments are sealed and indexes are all built as well. One quick question: since the index type was `IVF_FLAT`, what's the `nprobe` param value your used to search with? @xiaojuan8

**Reply by xiaojuan8** [reply]
Hi, search_params = {"metric_type": "L2", "params": {"nprobe": 10}, "offset": 5}
